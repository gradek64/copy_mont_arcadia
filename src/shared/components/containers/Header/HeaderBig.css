/** @define HeaderBig */

.HeaderBig {
  display: none;
  background-color: $header-big-background;
  max-width: $header-big-max-width;
  margin: $header-big-margin;
  padding: $header-big-padding;

  &-content {
    background-color: $header-background;
    height: $header-big-height;
    position: relative;
    width: 100%;
    max-width: $laptop-end-width;
    margin: 0 auto;
    display: flex;
    align-items: center;
    justify-content: center;
    padding-top: $header-big-content-padding-top;
  }

  &-brandLogo {
    height: $header-big-brand-logo-height;
  }

  &-centre {
    flex: $header-big-centre-flex-grow;
  }

  &-left {
    height: 100%;
    flex: $header-big-left-flex-grow;
    display: flex;
    align-items: $header-big-left-align-items;
    justify-content: $header-big-left-justify-content;
  }

  &-right {
    height: 100%;
    flex: $header-big-right-flex-grow;
    display: flex;
    align-items: $header-big-right-elements-alignment;
    justify-content: flex-end;
  }
}

@media $min-tablet {
  .HeaderBig {
    display: block;
    min-height: $header-big-min-height;
  }
}
