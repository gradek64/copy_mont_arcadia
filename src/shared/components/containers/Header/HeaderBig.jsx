/*
 * This is a dedicated component that we are going to use for all the widths bigger than 992.
 * We have chosen to create a brand new header component to make as less likely as possible the possibility to
 * break the mobile header.
 */

import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import * as headers from './'

@connect((state) => ({
  brandName: state.config.brandName,
  shoppingBagTotalItems: state.shoppingBag.totalItems,
  logoVersion: state.config.logoVersion,
  featureMegaNav: state.features.status.FEATURE_MEGA_NAV,
  region: state.config.region,
  sticky: state.pageHeader.sticky,
}))
class HeaderBig extends Component {
  static propTypes = {
    shoppingBagTotalItems: PropTypes.number.isRequired,
    brandName: PropTypes.string,
    logoVersion: PropTypes.string,
    featureMegaNav: PropTypes.bool,
    region: PropTypes.string,
    sticky: PropTypes.bool,
  }

  render() {
    const {
      region,
      brandName,
      shoppingBagTotalItems,
      logoVersion,
      featureMegaNav,
      sticky,
    } = this.props
    const HeaderComponent = headers[brandName]
    return (
      <div className="HeaderBig">
        <HeaderComponent
          sticky={sticky}
          brandName={brandName}
          contentWrapper="HeaderBig-content"
          leftCol="HeaderBig-left"
          rightCol="HeaderBig-right"
          centreCol="HeaderBig-centre"
          logoVersion={logoVersion}
          shoppingBagTotalItems={shoppingBagTotalItems}
          featureMegaNav={featureMegaNav}
          region={region}
        />
      </div>
    )
  }
}

export default HeaderBig
