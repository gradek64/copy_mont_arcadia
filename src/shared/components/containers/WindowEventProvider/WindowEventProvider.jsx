import { Component } from 'react'
import PropTypes from 'prop-types'

export class WindowEventProvider extends Component {
  static childContextTypes = {
    addListener: PropTypes.func,
    removeListener: PropTypes.func,
  }

  getChildContext() {
    const { addListener, removeListener } = this
    return {
      addListener,
      removeListener,
    }
  }

  addListener = (eventName, handler) => {
    console.log('add customEvent', eventName)
    console.log('customEvent handler', handler.name)
    window.addEventListener(eventName, handler)
  }

  removeListener = (eventName, handler) => {
    console.log('remove customEvent', eventName)
    console.log('customEvent handler', handler.name)
    window.removeEventListener(eventName, handler)
  }

  render() {
    return this.props.children
  }
}
