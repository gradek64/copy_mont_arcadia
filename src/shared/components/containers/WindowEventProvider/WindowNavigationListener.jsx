import { Component } from 'react'
import PropTypes from 'prop-types'

const POP_STATE_EVENT = 'popstate'

export class WindowNavigationListener extends Component {
  static contextTypes = {
    addListener: PropTypes.func,
    removeListener: PropTypes.func,
  }
  static propTypes = {
    modalOpen: PropTypes.bool,
    closeModal: PropTypes.func,
  }
  onBrowserNavigation = () => {
    const { modalOpen, closeModal } = this.props
    if (modalOpen && closeModal) closeModal()
    console.log(
      'POP_STATE_EVENT calling custom method event handler "onBrowserNavigation"'
    )
  }

  componentDidMount() {
    const { addListener } = this.context
    if (addListener) addListener(POP_STATE_EVENT, this.onBrowserNavigation)
  }

  componentWillUnmount() {
    const { removeListener } = this.context
    if (removeListener)
      removeListener(POP_STATE_EVENT, this.onBrowserNavigation)
  }

  render() {
    return this.props.children
  }
}
