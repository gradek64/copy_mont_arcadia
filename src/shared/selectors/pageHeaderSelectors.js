import { createSelector } from 'reselect'

import { isTablet, isPortrait } from './viewportSelectors'
import { getBrandCode, getLang } from './configSelectors'
import isForceMobileHeader from './isForceMobileHeader'

export const shouldDisplayMobileHeaderIfSticky = createSelector(
  isTablet,
  isPortrait,
  getLang,
  getBrandCode,
  (tablet, portrait, language, brandCode) =>
    isForceMobileHeader({
      isTablet: tablet,
      isPortrait: portrait,
      language,
      brandCode,
    })
)

const rootSelector = (state) => state.pageHeader || {}

export const isHeaderSticky = (state) => {
  const { sticky } = rootSelector(state)

  return Boolean(sticky)
}
