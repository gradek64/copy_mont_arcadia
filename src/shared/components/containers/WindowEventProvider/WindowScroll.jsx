import { Component } from 'react'
import PropTypes from 'prop-types'
import { WindowEventProvider } from './WindowEventProvider'
import throttle from 'lodash.throttle'

export class WindowScroll extends Component {
  constructor(props) {
    super(props)
    this.scrollHandler = throttle(this.handleScroll, props.scrollDelay)
  }

  static contextTypes = {
    ...WindowEventProvider.childContextTypes,
  }

  static propTypes = {
    children: PropTypes.element.isRequired,
    onScroll: PropTypes.func,
    onCustomScroll: PropTypes.func,
    onReachedPageBottom: PropTypes.func,
    scrollPastThreshold: PropTypes.number,
    pageBottomBuffer: PropTypes.number,
    scrollDelay: PropTypes.number.isRequired,
  }

  static defaultProps = {
    onScroll: () => undefined,
    onReachedPageBottom: () => undefined,
    scrollPastThreshold: 0,
    pageBottomBuffer: 0,
    scrollDelay: 300,
  }

  _hasPassedThreshold = false
  _hasReachedPageBottom = false

  handleScroll = (event, w = window) => {
    const { onScroll, onCustomScroll } = this.props

    if (onCustomScroll) {
      return onCustomScroll(event, w)
    }

    onScroll(event)
    this.checkScrollPassedThreshold(event, w)
    this.checkScrollReachedPageBottom(event, w)
  }

  checkScrollPassedThreshold(event, { innerHeight, scrollY, pageYOffset }) {
    const { onScrollPast, scrollPastThreshold } = this.props

    const scrollPosition = scrollY || pageYOffset

    if (scrollPastThreshold <= 0) return

    const heightThreshold =
      scrollPastThreshold <= 1
        ? innerHeight * scrollPastThreshold
        : scrollPastThreshold

    const hasPassedThreshold = scrollPosition > heightThreshold
    if (hasPassedThreshold !== this._hasPassedThreshold) {
      onScrollPast(hasPassedThreshold, event)
    }

    this._hasPassedThreshold = hasPassedThreshold
  }

  checkScrollReachedPageBottom(
    event,
    {
      innerHeight,
      scrollY,
      pageYOffset,
      document: {
        body: { offsetHeight },
      },
    }
  ) {
    const { onReachedPageBottom, pageBottomBuffer } = this.props

    const scrollPosition = scrollY || pageYOffset

    const hasReachedPageBottom =
      scrollPosition + innerHeight + pageBottomBuffer >= offsetHeight

    if (hasReachedPageBottom !== this._hasReachedPageBottom) {
      onReachedPageBottom(hasReachedPageBottom, event)
    }

    this._hasReachedPageBottom = hasReachedPageBottom
  }

  componentDidMount() {
    this.context.addListener('scroll', this.scrollHandler)
  }

  componentWillUnmount() {
    this.context.removeListener('scroll', this.scrollHandler)
  }

  render() {
    return this.props.children
  }
}
