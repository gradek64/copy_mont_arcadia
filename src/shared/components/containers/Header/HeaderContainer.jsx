import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import Header from './Header'
import HeaderBig from './HeaderBig'
import { WindowScroll } from '../WindowEventProvider/WindowScroll'
import { isMinViewPort } from '../../../selectors/viewportSelectors'
import { isFeatureStickyHeaderEnabled } from '../../../selectors/featureSelectors'
import { shouldDisplayMobileHeaderIfSticky } from '../../../selectors/pageHeaderSelectors'
import { isInCheckout } from '../../../selectors/routingSelectors'
import { setSticky } from '../../../actions/common/headerActions'
import { isCheckoutPath } from '../../../lib/checkout'

const stickyHeaderBreakpoints = {
  topshop: () => 38,
  topman: () => 38,
  missselfridge: () => 38,
  dorothyperkins: (isMinViewPort) => (isMinViewPort('laptop') ? 87 : 42),
  burton: () => 84,
  wallis: () => 0,
  evans: () => 0,
}

@connect(
  (state) => ({
    brandName: state.config.brandName,
    isMobile: state.viewport.media === 'mobile',
    featureBigHeader: state.features.status.FEATURE_HEADER_BIG,
    featureResponsive: state.features.status.FEATURE_RESPONSIVE,
    sticky: state.pageHeader.sticky,
    isInCheckout: isInCheckout(state),
    forceMobileHeaderIfSticky: shouldDisplayMobileHeaderIfSticky(state),
    isFeatureStickyHeaderEnabled: isFeatureStickyHeaderEnabled(state),
  }),
  {
    setSticky,
    isMinViewPort,
  }
)
class HeaderContainer extends Component {
  static propTypes = {
    brandName: PropTypes.string,
    location: PropTypes.object.isRequired,
    isMobile: PropTypes.bool.isRequired,
    featureBigHeader: PropTypes.bool,
    featureResponsive: PropTypes.bool,
    forceMobileHeaderIfSticky: PropTypes.bool,
    sticky: PropTypes.bool,
    isFeatureStickyHeaderEnabled: PropTypes.bool,
    isInCheckout: PropTypes.bool,
    isMinViewPort: PropTypes.func,
    isStickyMobile: PropTypes.bool,
  }

  constructor(props) {
    super(props)
    this.ref = React.createRef()
  }

  UNSAFE_componentWillReceiveProps(nextProps) {
    const { isInCheckout, sticky, setSticky, isMobile } = this.props
    const { isInCheckout: newIsInCheckout, isMobile: newIsMobile } = nextProps

    const isCheckoutStickyHeader = sticky && !isInCheckout && newIsInCheckout
    const isResizeIntoMobile = !isMobile && newIsMobile

    if (isCheckoutStickyHeader || isResizeIntoMobile) setSticky(false)
  }

  onScroll = (event, { scrollY, pageYOffset }) => {
    const y = scrollY || pageYOffset
    this.onUpdateScrollPosition(y)
  }

  onUpdateScrollPosition = (y) => {
    const {
      brandName,
      setSticky,
      sticky,
      isMinViewPort,
      isInCheckout,
    } = this.props
    if (!this.ref || !this.ref.current) return
    if (isInCheckout) return
    const offsetTop = this.ref.current.offsetTop

    const brandStickyOffset =
      stickyHeaderBreakpoints[brandName](isMinViewPort) || 0
    const stickyOffsetBreached = y > offsetTop + brandStickyOffset
    const newSticky = stickyOffsetBreached && !isInCheckout
    if ((newSticky && !sticky) || (!newSticky && sticky)) setSticky(newSticky)
  }

  renderHeader = () => {
    const {
      isMobile,
      location: { pathname },
      featureBigHeader,
      featureResponsive,
      forceMobileHeaderIfSticky,
      sticky,
      isStickyMobile,
    } = this.props
    const isCheckout = isCheckoutPath(pathname)
    const isCheckoutLogin = isCheckout && pathname.includes('login')
    const isCheckoutSummary = isCheckout && pathname.includes('summary')
    const isCheckoutBig = isCheckout && featureBigHeader && !isMobile

    const bigHeader =
      featureBigHeader && featureResponsive && !isMobile && !isCheckout

    const forceMobile = forceMobileHeaderIfSticky && sticky
    return (
      <React.Fragment>
        <Header
          hasMenuButton={!isCheckout}
          hasSearchBar={
            !isCheckout && (isMobile || !featureResponsive || !featureBigHeader)
          }
          hasCartIcon={!isCheckoutLogin && !isCheckoutSummary && !isCheckoutBig}
          isCheckoutBig={isCheckoutBig}
          forceDisplay={forceMobile}
          isStickyMobile={isStickyMobile}
          ref={this.ref}
        />
        {bigHeader && !forceMobile && <HeaderBig />}
      </React.Fragment>
    )
  }

  render() {
    const { isFeatureStickyHeaderEnabled, isMobile } = this.props
    const scrollDelay = 50

    if (isFeatureStickyHeaderEnabled && !isMobile) {
      return (
        <WindowScroll scrollDelay={scrollDelay} onCustomScroll={this.onScroll}>
          {this.renderHeader()}
        </WindowScroll>
      )
    }
    return this.renderHeader()
  }
}

export default HeaderContainer
