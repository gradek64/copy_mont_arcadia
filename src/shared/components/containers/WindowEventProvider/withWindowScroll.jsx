import React from 'react'
import { WindowScroll } from './WindowScroll'

export const withWindowScroll = ({
  scrollPastThreshold = 0,
  notifyWhenReachedBottomOfPage = false,
  pageBottomBuffer = 0,
}) => (WrappedComponent) => {
  class component extends React.Component {
    state = {
      hasPassedThreshold: false,
      hasReachedPageBottom: false,
    }

    onScrollPast = (hasPassedThreshold) => {
      this.setState({
        hasPassedThreshold,
      })
    }

    onReachedPageBottom = (hasReachedPageBottom) => {
      if (notifyWhenReachedBottomOfPage) {
        this.setState({
          hasReachedPageBottom,
        })
      }
    }

    render() {
      return (
        <WindowScroll
          scrollPastThreshold={scrollPastThreshold}
          onScrollPast={this.onScrollPast}
          onReachedPageBottom={this.onReachedPageBottom}
          pageBottomBuffer={pageBottomBuffer}
        >
          <WrappedComponent
            {...this.props}
            hasPassedThreshold={this.state.hasPassedThreshold}
            hasReachedPageBottom={this.state.hasReachedPageBottom}
          />
        </WindowScroll>
      )
    }
  }

  component.WrappedComponent = WrappedComponent
  return component
}
