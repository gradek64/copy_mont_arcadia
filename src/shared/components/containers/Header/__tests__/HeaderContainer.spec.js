import { shallow } from 'enzyme'
import React from 'react'
import testComponentHelper from 'test/unit/helpers/test-component'
import HeaderContainer from '../HeaderContainer'
import HeaderBig from '../HeaderBig'
import configureStore from 'redux-mock-store'
import { isMinViewPort } from '../../../../selectors/viewportSelectors'
import { shouldDisplayMobileHeaderIfSticky } from '../../../../selectors/pageHeaderSelectors'
import { WindowScroll } from '../../WindowEventProvider/WindowScroll'

jest.mock('../../../../selectors/viewportSelectors', () => ({
  isMinViewPort: jest.fn().mockReturnValue(() => false),
  isMobile: jest.fn().mockReturnValue(() => false),
}))
jest.mock('../../../../selectors/pageHeaderSelectors', () => ({
  shouldDisplayMobileHeaderIfSticky: jest.fn().mockReturnValue(false),
}))

const brandName = 'topshop'

const setSticky = jest.fn()

beforeEach(() => jest.clearAllMocks())

describe('<HeaderContainer />', () => {
  const initProps = {
    location: {
      pathname: '/product/',
    },
    isMinViewPort,
    isMobile: true,
    brandName: 'topshop',
    forceMobileHeaderIfSticky: false,
    sticky: false,
    setSticky,
    isStickyMobile: false,
  }
  const isResponsive = {
    featureResponsive: true,
    featureBigHeader: true,
  }

  describe('Connected component', () => {
    it('should receive the correct props', () => {
      const featureBigHeader = true
      const featureResponsive = true
      const initialState = {
        viewport: {
          media: 'mobile',
        },
        features: {
          status: {
            FEATURE_HEADER_BIG: featureBigHeader,
            FEATURE_RESPONSIVE: featureResponsive,
          },
        },
        pageHeader: {
          sticky: true,
        },
        config: {
          brandName,
        },
      }
      expect(shouldDisplayMobileHeaderIfSticky).toHaveBeenCalledTimes(0)
      const container = shallow(
        <HeaderContainer location={{}} store={configureStore()(initialState)} />
      )
      expect(shouldDisplayMobileHeaderIfSticky).toHaveBeenCalledTimes(1)
      expect(container).toBeTruthy()
      expect(container.prop('isMobile')).toBeTruthy()
      expect(container.prop('sticky')).toBe(true)
      expect(container.prop('brandName')).toBe(brandName)
      expect(container.prop('featureBigHeader')).toBe(featureBigHeader)
      expect(container.prop('featureResponsive')).toBe(featureResponsive)
      expect(container.prop('forceMobileHeaderIfSticky')).toBe(false)
      expect(container.prop('isMinViewPort')).toEqual(expect.any(Function))
    })
  })

  const renderComponent = testComponentHelper(HeaderContainer.WrappedComponent)
  describe('@renders', () => {
    describe('@mobile', () => {
      beforeEach(() => jest.clearAllMocks())

      it('in default state', () => {
        expect(renderComponent(initProps).getTree()).toMatchSnapshot()
      })
      it('does not render HeaderBig', () => {
        const { wrapper } = renderComponent(initProps)
        expect(wrapper.find(HeaderBig)).toHaveLength(0)
      })
      it('does not forceDisplay header by default', () => {
        const { wrapper } = renderComponent(initProps)
        expect(wrapper.find('Header').prop('forceDisplay')).toBe(false)
      })
      it('does not forceDisplay if sticky=true', () => {
        const { wrapper } = renderComponent({ ...initProps, sticky: true })
        expect(wrapper.find('Header').prop('forceDisplay')).toBe(false)
      })
      it('does not forceDisplay if forceMobileHeaderIfSticky=true', () => {
        const { wrapper } = renderComponent({
          ...initProps,
          forceMobileHeaderIfSticky: true,
        })
        expect(wrapper.find('Header').prop('forceDisplay')).toBe(false)
      })
      it('sets forceDisplay if forceMobileHeaderIfSticky=true', () => {
        const { wrapper } = renderComponent({
          ...initProps,
          sticky: true,
          forceMobileHeaderIfSticky: true,
        })
        expect(wrapper.find('Header').prop('forceDisplay')).toBe(true)
      })
      it('in login page', () => {
        expect(
          renderComponent({
            ...initProps,
            location: {
              pathname: '/login/',
            },
          }).getTree()
        ).toMatchSnapshot()
      })
      it('in summary page', () => {
        expect(
          renderComponent({
            ...initProps,
            location: {
              pathname: '/summary/',
            },
          }).getTree()
        ).toMatchSnapshot()
      })
      it('in checkout page', () => {
        expect(
          renderComponent({
            ...initProps,
            location: {
              pathname: '/checkout/',
            },
          }).getTree()
        ).toMatchSnapshot()
      })
      it('in checkout page and featureResponsive is enabled', () => {
        expect(
          renderComponent({
            ...initProps,
            ...isResponsive,
            location: {
              pathname: '/checkout/',
            },
          }).getTree()
        ).toMatchSnapshot()
      })
      it('in product page and featureResponsive is enabled', () => {
        expect(
          renderComponent({
            ...initProps,
            ...isResponsive,
          }).getTree()
        ).toMatchSnapshot()
      })
    })
    describe('@desktop', () => {
      it('in default state', () => {
        expect(
          renderComponent({ ...initProps, isMobile: false }).getTree()
        ).toMatchSnapshot()
      })
      it('does not render HeaderBig by default', () => {
        const { wrapper } = renderComponent({
          ...initProps,
          isMobile: false,
        })
        expect(wrapper.find(HeaderBig)).toHaveLength(0)
      })
      it('renders HeaderBig when responsive', () => {
        const { wrapper } = renderComponent({
          ...initProps,
          ...isResponsive,
          isMobile: false,
        })
        expect(wrapper.find(HeaderBig)).toHaveLength(1)
      })
      it('renders HeaderBig when sticky', () => {
        const { wrapper } = renderComponent({
          ...initProps,
          ...isResponsive,
          sticky: true,
          isMobile: false,
        })
        expect(wrapper.find(HeaderBig)).toHaveLength(1)
      })
      it('renders HeaderBig when forceMobileHeaderIfSticky=true', () => {
        const { wrapper } = renderComponent({
          ...initProps,
          ...isResponsive,
          forceMobileHeaderIfSticky: true,
          isMobile: false,
        })
        expect(wrapper.find(HeaderBig)).toHaveLength(1)
      })
      it('does not render HeaderBig when forceMobileHeaderIfSticky=true and sticky=true', () => {
        const { wrapper } = renderComponent({
          ...initProps,
          ...isResponsive,
          forceMobileHeaderIfSticky: true,
          sticky: true,
          isMobile: false,
        })
        expect(wrapper.find(HeaderBig)).toHaveLength(0)
      })
      it('in checkout page', () => {
        expect(
          renderComponent({
            ...initProps,
            location: {
              pathname: '/checkout/',
            },
            isMobile: false,
          }).getTree()
        ).toMatchSnapshot()
      })
      it('in product page and featureResponsive is enabled', () => {
        expect(
          renderComponent({
            ...initProps,
            ...isResponsive,
            isMobile: false,
          }).getTree()
        ).toMatchSnapshot()
      })
      it('in checkout page and featureResponsive is enabled', () => {
        expect(
          renderComponent({
            ...initProps,
            ...isResponsive,
            location: {
              pathname: '/checkout/',
            },
            isMobile: false,
          }).getTree()
        ).toMatchSnapshot()
      })
      it('wraps with WindowScroll when isMobile=false', () => {
        const { wrapper } = renderComponent({
          ...initProps,
          ...isResponsive,
          isFeatureStickyHeaderEnabled: true,
          isMobile: false,
        })
        expect(wrapper.find(WindowScroll)).toHaveLength(1)
      })
      it('does not wrap with WindowScroll when isMobile=true', () => {
        const { wrapper } = renderComponent({
          ...initProps,
          ...isResponsive,
          isFeatureStickyHeaderEnabled: true,
          isMobile: true,
        })
        expect(wrapper.find(WindowScroll).exists()).toEqual(false)
      })
    })
  })

  describe('@lifecycle', () => {
    describe('UNSAFE_componentWillReceiveProps', () => {
      it('calls setSticky(false) if sticky and new location is checkout', () => {
        const { instance, wrapper } = renderComponent({
          ...initProps,
          isInCheckout: false,
          sticky: true,
        })
        wrapper.update()
        instance.UNSAFE_componentWillReceiveProps({
          ...initProps,
          isInCheckout: true,
        })
        wrapper.update()
        expect(setSticky).toHaveBeenCalledTimes(1)
        expect(setSticky).toHaveBeenLastCalledWith(false)
      })
      it('calls setSticky(false) if new layout is mobile and current layout is not', () => {
        const { instance, wrapper } = renderComponent({
          ...initProps,
          isMobile: false,
        })
        wrapper.update()
        instance.UNSAFE_componentWillReceiveProps({
          ...initProps,
          isMobile: true,
        })
        wrapper.update()
        expect(setSticky).toHaveBeenCalledTimes(1)
        expect(setSticky).toHaveBeenLastCalledWith(false)
      })
      it('does not call setSticky if not sticky and new location is checkout and new layout is not mobile', () => {
        const { instance, wrapper } = renderComponent({
          ...initProps,
          isInCheckout: false,
          sticky: false,
          isMobile: true,
        })
        wrapper.update()
        instance.UNSAFE_componentWillReceiveProps({
          ...initProps,
          isInCheckout: true,
          isMobile: false,
        })
        wrapper.update()
        expect(setSticky).not.toHaveBeenCalled()
      })
    })
  })

  describe('@events', () => {
    beforeEach(() => {
      setSticky.mockReset()
      isMinViewPort.mockReset()
    })
    it('calls setSticky action on first scroll movement around y position', () => {
      const { wrapper } = renderComponent({
        ...initProps,
        isFeatureStickyHeaderEnabled: true,
      })
      const scrollPoint = 38
      wrapper.instance().ref = { current: { offsetTop: 0 } }
      expect(setSticky).toHaveBeenCalledTimes(0)
      wrapper.instance().onUpdateScrollPosition(scrollPoint)
      expect(setSticky).toHaveBeenCalledTimes(0)
      wrapper.instance().onUpdateScrollPosition(scrollPoint + 1)
      expect(setSticky).toHaveBeenLastCalledWith(true)
      wrapper.setProps({ sticky: true })
      setSticky.mockReset()
      wrapper.instance().onUpdateScrollPosition(scrollPoint + 2)
      expect(setSticky).toHaveBeenCalledTimes(0)
      wrapper.instance().onUpdateScrollPosition(scrollPoint + 1)
      expect(setSticky).toHaveBeenCalledTimes(0)
      wrapper.instance().onUpdateScrollPosition(scrollPoint)
      expect(setSticky).toHaveBeenLastCalledWith(false)
      wrapper.setProps({ sticky: false })
      setSticky.mockReset()
      wrapper.instance().onUpdateScrollPosition(scrollPoint - 1)
      expect(setSticky).toHaveBeenCalledTimes(0)
    })
    it('does not call setSticky action if isInCheckout is true', () => {
      const { wrapper } = renderComponent({
        ...initProps,
        isFeatureStickyHeaderEnabled: true,
        isInCheckout: true,
      })
      const scrollPoint = 38
      wrapper.instance().ref = { current: { offsetTop: 0 } }
      wrapper.instance().onUpdateScrollPosition(scrollPoint + 50)
      expect(setSticky).toHaveBeenCalledTimes(0)
    })
    it('does not call setSticky action if ref is not defined', () => {
      const { wrapper } = renderComponent({
        ...initProps,
        isFeatureStickyHeaderEnabled: true,
      })
      const scrollPoint = 38
      wrapper.instance().ref = null
      wrapper.instance().onUpdateScrollPosition(scrollPoint + 50)
      expect(setSticky).toHaveBeenCalledTimes(0)
    })
    it('dorothy perkins invokes sticky at breakpoint for >= laptop viewport', () => {
      const { wrapper } = renderComponent({
        ...initProps,
        isFeatureStickyHeaderEnabled: true,
        brandName: 'dorothyperkins',
      })
      isMinViewPort.mockImplementation(() => true)
      const scrollPoint = 87
      wrapper.instance().ref = { current: { offsetTop: 0 } }
      expect(setSticky).toHaveBeenCalledTimes(0)
      wrapper.instance().onUpdateScrollPosition(scrollPoint)
      expect(setSticky).toHaveBeenCalledTimes(0)
      wrapper.instance().onUpdateScrollPosition(scrollPoint + 1)
      expect(setSticky).toHaveBeenLastCalledWith(true)
      expect(isMinViewPort).toHaveBeenCalledTimes(2)
    })
    it('dorothy perkins invokes sticky at breakpoint for < laptop viewport', () => {
      const { wrapper } = renderComponent({
        ...initProps,
        isFeatureStickyHeaderEnabled: true,
        brandName: 'dorothyperkins',
      })
      isMinViewPort.mockImplementation(() => false)
      const scrollPoint = 42
      wrapper.instance().ref = { current: { offsetTop: 0 } }
      expect(setSticky).toHaveBeenCalledTimes(0)
      wrapper.instance().onUpdateScrollPosition(scrollPoint)
      expect(setSticky).toHaveBeenCalledTimes(0)
      wrapper.instance().onUpdateScrollPosition(scrollPoint + 1)
      expect(setSticky).toHaveBeenLastCalledWith(true)
      expect(isMinViewPort).toHaveBeenCalledTimes(2)
    })
  })

  describe('onScroll', () => {
    it('should call onUpdateScrollPosition on window scroll', () => {
      const mockEvent = { event: 'mockEvent' }
      const mockWindow = {
        scrollY: 10,
        pageYOffset: 10,
      }
      const { instance } = renderComponent(initProps)
      const onUpdateScrollPositionMock = jest.spyOn(
        instance,
        'onUpdateScrollPosition'
      )

      expect(onUpdateScrollPositionMock).toHaveBeenCalledTimes(0)
      instance.onScroll(mockEvent, mockWindow)
      expect(onUpdateScrollPositionMock).toHaveBeenCalledTimes(1)
      expect(onUpdateScrollPositionMock).toHaveBeenCalledWith(10)
    })
  })
})
