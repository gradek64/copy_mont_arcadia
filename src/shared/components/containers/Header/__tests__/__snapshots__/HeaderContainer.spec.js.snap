// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HeaderContainer /> @renders @desktop in checkout page 1`] = `
<Fragment>
  <Header
    forceDisplay={false}
    hasCartIcon={true}
    hasMenuButton={false}
    hasSearchBar={false}
    isStickyMobile={false}
  />
</Fragment>
`;

exports[`<HeaderContainer /> @renders @desktop in checkout page and featureResponsive is enabled 1`] = `
<Fragment>
  <Header
    forceDisplay={false}
    hasCartIcon={false}
    hasMenuButton={false}
    hasSearchBar={false}
    isCheckoutBig={true}
    isStickyMobile={false}
  />
</Fragment>
`;

exports[`<HeaderContainer /> @renders @desktop in default state 1`] = `
<Fragment>
  <Header
    forceDisplay={false}
    hasCartIcon={true}
    hasMenuButton={true}
    hasSearchBar={true}
    isCheckoutBig={false}
    isStickyMobile={false}
  />
</Fragment>
`;

exports[`<HeaderContainer /> @renders @desktop in product page and featureResponsive is enabled 1`] = `
<Fragment>
  <Header
    forceDisplay={false}
    hasCartIcon={true}
    hasMenuButton={true}
    hasSearchBar={false}
    isCheckoutBig={false}
    isStickyMobile={false}
  />
  <Connect(HeaderBig) />
</Fragment>
`;

exports[`<HeaderContainer /> @renders @mobile in checkout page 1`] = `
<Fragment>
  <Header
    forceDisplay={false}
    hasCartIcon={true}
    hasMenuButton={false}
    hasSearchBar={false}
    isStickyMobile={false}
  />
</Fragment>
`;

exports[`<HeaderContainer /> @renders @mobile in checkout page and featureResponsive is enabled 1`] = `
<Fragment>
  <Header
    forceDisplay={false}
    hasCartIcon={true}
    hasMenuButton={false}
    hasSearchBar={false}
    isCheckoutBig={false}
    isStickyMobile={false}
  />
</Fragment>
`;

exports[`<HeaderContainer /> @renders @mobile in default state 1`] = `
<Fragment>
  <Header
    forceDisplay={false}
    hasCartIcon={true}
    hasMenuButton={true}
    hasSearchBar={true}
    isCheckoutBig={false}
    isStickyMobile={false}
  />
</Fragment>
`;

exports[`<HeaderContainer /> @renders @mobile in login page 1`] = `
<Fragment>
  <Header
    forceDisplay={false}
    hasCartIcon={true}
    hasMenuButton={true}
    hasSearchBar={true}
    isCheckoutBig={false}
    isStickyMobile={false}
  />
</Fragment>
`;

exports[`<HeaderContainer /> @renders @mobile in product page and featureResponsive is enabled 1`] = `
<Fragment>
  <Header
    forceDisplay={false}
    hasCartIcon={true}
    hasMenuButton={true}
    hasSearchBar={true}
    isCheckoutBig={false}
    isStickyMobile={false}
  />
</Fragment>
`;

exports[`<HeaderContainer /> @renders @mobile in summary page 1`] = `
<Fragment>
  <Header
    forceDisplay={false}
    hasCartIcon={true}
    hasMenuButton={true}
    hasSearchBar={true}
    isCheckoutBig={false}
    isStickyMobile={false}
  />
</Fragment>
`;
